examples/ex71.hpp

```diff
@@ -0,0 +1,562 @@
+      ir = &IntRules.Get(el.GetGeomType(), order);
+
+      Vector shapef(ndof);
+      DenseMatrix dshape_iso(ndof, ndim);
+      DenseMatrix dshape_xyz(ndof, spaceDim);
+      elmat.SetSize(ndof, ndof);
+      elmat = 0.0;
+
+      DenseMatrix B(ndof, 4); // [diff_x,diff_y,diff_z, shape]
+      DenseMatrix A(ndof, 4);
+      Vector vparam(3); // [power, epsilon, load]
+      Vector uu(4);     // [diff_x,diff_y,diff_z,u]
+      DenseMatrix duu(4, 4);
+      B = 0.0;
+      uu = 0.0;
+      // initialize the parameters - keep the same order
+      // utilized in the pLapIntegrator definition
+      vparam[0] = 2.0;  // default power
+      vparam[1] = 1e-8; // default epsilon
+      vparam[2] = 1.0;  // default load
+
+      double w;
+
+      for (int i = 0; i < ir->GetNPoints(); i++)
+      {
+         const IntegrationPoint &ip = ir->IntPoint(i);
+         trans.SetIntPoint(&ip);
+         w = trans.Weight();
+         w = ip.weight * w;
+
+         el.CalcDShape(ip, dshape_iso);
+         el.CalcShape(ip, shapef);
+         Mult(dshape_iso, trans.InverseJacobian(), dshape_xyz);
+
+         // set the matrix B
+         for (int jj = 0; jj < spaceDim; jj++)
+         {
+            B.SetCol(jj, dshape_xyz.GetColumn(jj));
+         }
+         B.SetCol(3, shapef);
+
+         // set the power
+         if (pp != nullptr)
+         {
+            vparam[0] = pp->Eval(trans, ip);
+         }
+         // set the coefficient ensuring positiveness of the tangent matrix
+         if (coeff != nullptr)
+         {
+            vparam[1] = coeff->Eval(trans, ip);
+         }
+         // add the contribution from the load
+         if (load != nullptr)
+         {
+            vparam[2] = load->Eval(trans, ip);
+         }
+
+         // calculate uu
+         B.MultTranspose(elfun, uu);
+         // calculate derivative of the energy with respect to uu
+         qint.QFunctionDD(vparam, uu, duu);
+
+         Mult(B, duu, A);
+         AddMult_a_ABt(w, A, B, elmat);
+
+      } // end integration loop
+   }
+};
+
+class pLaplace : public NonlinearFormIntegrator
+{
+protected:
+   Coefficient *pp;
+   Coefficient *coeff;
+   Coefficient *load;
+
+public:
+   pLaplace()
+   {
+      coeff = nullptr;
+      pp = nullptr;
+   }
+
+   pLaplace(Coefficient &pp_) : pp(&pp_), coeff(nullptr), load(nullptr) {}
+
+   pLaplace(Coefficient &pp_, Coefficient &q, Coefficient &ld_)
+      : pp(&pp_), coeff(&q), load(&ld_)
+   {}
+
+   virtual ~pLaplace() {}
+
+   virtual double GetElementEnergy(const FiniteElement &el,
+                                   ElementTransformation &trans,
+                                   const Vector &elfun) override
+   {
+      double energy = 0.0;
+      int ndof = el.GetDof();
+      int ndim = el.GetDim();
+      int spaceDim = trans.GetSpaceDim();
+      bool square = (ndim == spaceDim);
+      const IntegrationRule *ir = NULL;
+      int order = 2 * trans.OrderGrad(&el) - 1; // correct order?
```

Comment:
Is it the correct order? Check and remove comment.

Reply:
corrected